### ISSUE TITLE
Deprecate Class::getData('variable') idioms in favour of Class::getVariable() functions

### ISSUE THREAD

**Issue by johnsaigle** [aces/Loris#5358]
Many LORIS classes are structured in a weird way.

E.g. The TimePoint class has information stored in a private array:

```php
private var $_timePointInfo = array();
```

This array gets populates when the object gets instantiated. It contains numerous variables about the TimePoint such as its center ID, subprojectID, etc.

There exists a function

```php
public function getData(?string $key = null);
```

Which can be used to get the value of these variables. At the same time, there exist individual functions to do this:

```php
public function getCenterID();
```
```php
public function getSubprojectID;
```

---

## So what?

* There's no point having a private array of variables that can be accessed publicly. They're not really private in this case
* Using `TimePoint::getData()` prevents type hinting since this function needs to return `?mixed` as it is a catch-all for all data the object uses
* Creating an array of what are essentially class variables defeats the whole point of having a class. You might as well use arrays everywhere because this basically reduces the class to being an anonymous struct with no typing
* Without class variables, we will be unable to take advantage of class variable typing in PHP 7.4
* All of these points are likely to make things more error-prone as scope isn't respected and types can't be used.

## Solution

Refactor TimePoint and other classes that are designed this way to have EITHER

1. Public class variables OR
2. Private class variables with traditional getters/setters

Additionally, new code going into LORIS should use something like `TimePoint::getCenterID()` rather than `TimePoint::getData('centerID')` as this will allow us to make use of type hinting on functions of the latter type as well as smooth the transition to better classes.

**Comment by johnsaigle** [comment]
It may also be a good idea to put a DeprecationException in `Class::getData()` functions informing developers to use more specific functions instead.

**Comment by driusan** [comment]
>    There's no point having a private array of variables that can be accessed publicly. They're not really private in this case

That's not really true. It prevents setting and allows the possibility of modifying the data in some way before returning it.

>    Using TimePoint::getData() prevents type hinting since this function needs to return ?mixed as it is a catch-all for all data the object uses

I think the easiest solution is simply to make it protected and use it internally in the getters, which can do their own casting. 

>    Creating an array of what are essentially class variables defeats the whole point of having a class. You might as well use arrays everywhere because this basically reduces the class to being an anonymous struct with no typing

Since it's private and people need to use the getter, this is just an implementation detail. I suspect it's done this way so that it can be easily retrieved from the database and put into the class.. (as an aside: arrays and objects are just different syntax on the same internal structure in PHP.)

**Comment by johnsaigle** [comment]
I'm adding Bug to this because I feel that this is important and don't want to fight Stale bot.

**Comment by driusan** [comment]
I agree that this shouldn't be closed, but I don't think the "Bug" label is appropriate here.. and at any rate, the worst case scenerio is that it gets a "stale" tag with the current configuration if there's no activity (which is probably an accurate description of the issue's state)
